source/applications/advanced/color_balance.py

```diff
@@ -0,0 +1,257 @@
+"""
+Balance color for RGB image.
+"""
+
+import datetime
+from pathlib import Path
+import numpy as np
+import matplotlib.pyplot as plt
+import zivid
+
+
+def _display_rgb(rgb, title):
+    """Display RGB image.
+
+    Args:
+        rgb: RGB image
+        title: Image title
+
+    Returns None
+
+    """
+    plt.figure()
+    plt.imshow(rgb)
+    plt.title(title)
+    plt.show(block=False)
+
+
+def _compute_mean_rgb(rgb, pixels):
+    """Compute mean RGB values.
+
+    Args:
+        rgb: RGB image
+        pixels: Number of central pixels for computation
+
+    Returns:
+        mean_red: Red channel mean value
+        mean_green: Green channel mean value
+        mean_blue: Blue channel mean value
+
+    """
+    height = np.shape(rgb)[0]
+    width = np.shape(rgb)[1]
+    red = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        0,
+    ]
+    green = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        1,
+    ]
+    blue = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        2,
+    ]
+
+    mean_red = np.mean(np.reshape(red, -1, []))
+    mean_green = np.mean(np.reshape(green, -1, []))
+    mean_blue = np.mean(np.reshape(blue, -1, []))
+
+    return mean_red, mean_green, mean_blue
+
+
+def _rgb_from_3d_capture(camera):
+    """Get RGB image from 3D capture.
+
+    Args:
+        camera: Zivid camera
+
+    Returns:
+        rgb: RGB image
+
+    """
+    frame = camera.capture()
+    point_cloud = frame.get_point_cloud().to_array()
+    rgb = np.dstack([point_cloud["r"], point_cloud["g"], point_cloud["b"]])
+
+    return rgb
+
+
+def _color_balance_calibration(camera):
+    """Balance color for RGB image by taking images of white surface (piece of paper, wall, etc.) in a loop.
+
+    Args:
+        camera: Zivid camera
+
+    Returns:
+        corrected_red_balance: Corrected red balance
+        corrected_blue_balance: Corrected blue balance
+
+    """
+    print(f"Starting color balance calibration")
+    default_red_balance = camera.settings.red_balance
+    default_blue_balance = camera.settings.blue_balance
+    with camera.update_settings() as updater:
+        updater.settings.blue_balance = 1.0
+        updater.settings.red_balance = 1.0
+
+    rgb = _rgb_from_3d_capture(camera)
```

Comment:
I guess it should be possible to do color calibration for 2D as well? I  understand that this works well due to being able to set rgain and bgain in 3D. But wasn't the main problem 2D?

Reply:
The problem is that a few iterations are necessary (most times 3) to get rgain and bgain that make R G B mean values equal. That is why the calibration is in 3D (to get those rgain and bgain) and correction is in 2D (where you apply rgain and bgain on the image).

Reply:
I haven't changed this. In average, three iterations are required.